docs/design-proposals/2022-04 TraceQL Concepts.md

```diff
@@ -0,0 +1,203 @@
+
+This document is **not** meant to be a complete language specification. Rather it is meant to invite the community to comment on the basic structure and concepts of the language.
+
+### Table of Contents
+
+- [Capabilities](#capabilities)
+- [Structure](#structure)
+- [Selecting spans](#selecting-spans)
+  - [Intrinsic fields](#intrinsic-fields)
+  - [Attribute fields](#attribute-fields)
+  - [Field expressions](#field-expressions)
+- [Combining Spansets](#combining-spansets)
+  - [Structural Operators](#structural-operators)
+- [Aggregators](#aggregators)
+- [Pipelining](#pipelining)
+- [Grouping](#grouping)
+- [Examples](#examples)
+  - [Shorthand](#shorthand)
+
+## Capabilities
+
+A TraceQL query can select traces based on 
+
+- span attributes, timing and duration
+- structural relationships between spans
+- aggregated data from the spans in a trace
+
+The TraceQL language is designed to use the same syntax and semantics as PromQL and LogQL where possible. The syntax and semantics differ due to the specialized needs of querying traces.
+
+## Structure
+
+A query is an expression that is evaluated on one trace at a time. The query is structured as a set of chained
+expressions (a pipeline). Each expression in the pipeline selects or discards spansets from being included in the results set. E.g.
+
+`{ http.status = 200 } | by(namespace) | count() > 3`
+
+Queries select sets of spans and filter them through a pipeline of aggregators and conditions. If a spanset is produced after evaluation
+on a trace then this spanset (and by extension the trace) is included in the resultset of the query.
+
+## Selecting spans
+
+In TraceQL the curly brackets `{}` always select a set of spans from the current trace. They will commonly be paired with a condition to reduce the spans being passed in.
+
+`{ http.status = 200 }`
+
+This simple query will be evaluated on every span of every trace, one at a time. 
+
+If the trace being evaluated contains no spans with an attribute `http.status` with the value `200` then no spans will be selected and this trace will not appear in our resultset.
+
+If the trace does contain spans with an attribute `http.status` with the value `200` then only those spans will be returned. The trace is reduced to only the set of spans that match the condition inside the `{}`. This set of spans (and therefore trace) will be included in our resultset.
+
+### Intrinsic fields
+
+Every span has certain intrinsic fields that can be referenced when selecting spans
+
+| field name | description |
+| ---------- | ----------- |
+| `start`    | start time of the span |
+| `end`      | end time of the span |
+| `duration` | end - start time of the span |
+| `name`     | operation or span name |
+| `status`   | status values are "error", "ok", or "unset" |
+| `parent`   | the parent of this span | 
```

Comment:
The `parent` intrinsic I feel has some hidden complexity and wonder if it is a worthwhile element of the language.  A basic spanset operator such as `{ A && B}` is evaluated on an individual span which is a straightforward concept and allows for simple storage mechanisms and algorithms.  However `{ parent.A && B }` evaluates across multiple spans.  The implied downstream implementation must now theoretically have both spans present in-memory and evaluated at the same time.  For a worst case example consider the possibility of `parent.parent.parent...`.


The `parent` intrinsic could be replaced with the existing child operator:  `{ parent.A && B }` is equal to ` {A} > {B}`, which is more consistent with the rest of the language, and also the implementation is clear.

Reply:
I do like parent for expressions such as:
`{ parent.duration - duration > 1s }` 
or 
`{ parent.service.name != service.name }`

Also, I"m fine with restricting parent to only being allowed to have depth of 1. Let's dig into the details as we work toward an internal POC. This is still an evolving spec and we can make adjustments.

Reply:
@mdisibio your complexity argument doesn't sit well with me. The whole point of a dedicated trace QL to me is its ability to express queries across spans of the trace; if it didn't do that I could just as well stick with SQL. So any sensible implementation of TraceQL would have to deal with conditions not defined on individual spans, this complexity is unavoidable.

The language purity argument makes sense, I think it's a matter of how difficult it is to express the same clauses that Joe showed above using the `{A} > {B}` notation.